### PR TITLE
improvement: reconstruct `type: date/datetime` to simplify date type

### LINKED ISSUE #4912: Simplify date/datetime datatypes
### Discussed in https://github.com/datafuselabs/databend/discussions/4909

<div type='discussions-op-text'>

<sup>Originally posted by **sundy-li** April 18, 2022</sup>
Proposal:

1.  Union `Date`(`Date16`)    `Date32` to be `Date` type, which uses i32 as physical datatype.
2.   Union `DateTime` (`DateTime32`), `DateTime64` to be `DateTime` type, which uses i64 as physical datatype.</div>

### PR DESCRIPTION
**Pull request by Veeupup**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

reconstruct `date` `datetime` datatype to simplify date type, discussion in #4909 
and fix a little bug in metactl export, make export date sorted.

TODO(in next pr):
* rename `datetime` type with new name and modify parser to support `DateTime(3)/DateTime(6)/DateTime(9)` datatype

## Changelog

- Improvement

## Related Issues

Fixes #4912

### CODE REVIEW

File: common/functions/src/scalars/others/running_difference_function.rs [hunk 1]
```diff
@@ -38,14 +38,12 @@ impl RunningDifferenceFunction {
 
         let output_type = match dt.data_type_id() {
             TypeID::Int8 | TypeID::UInt8 => Ok(type_primitive::Int16Type::arc()),
-            TypeID::Int16 | TypeID::UInt16 | TypeID::Date16 => Ok(type_primitive::Int32Type::arc()),
+            TypeID::Int16 | TypeID::UInt16 | TypeID::Date => Ok(type_primitive::Int32Type::arc()),
```
Comment (sundy-li) [inline common/functions/src/scalars/others/running_difference_function.rs:41]: Wrong!

File: common/functions/src/scalars/others/running_difference_function.rs [hunk 2]
```diff
@@ -38,14 +38,12 @@ impl RunningDifferenceFunction {
...
             TypeID::Int32
             | TypeID::UInt32
             | TypeID::Int64
             | TypeID::UInt64
-            | TypeID::Date32
-            | TypeID::DateTime32
-            | TypeID::DateTime64
+            | TypeID::DateTime
```
Comment (sundy-li) [inline common/functions/src/scalars/others/running_difference_function.rs:46]: |  TypeID::Date | TypeID::DateTime

Comment (Veeupup) [reply common/functions/src/scalars/others/running_difference_function.rs:46]: get

Comment (Veeupup) [reply common/functions/src/scalars/others/running_difference_function.rs:47]: resoloved

**Comment by Veeupup** [comment]
@dantengsky  @sundy-li  it seems that we encounter the compatible problem with `meta service` and `fuse engine`.Because meta service and fuse engine have persisted the `type` info.

File: common/functions/src/scalars/expressions/cast_from_string.rs [hunk 3]
```diff
@@ -51,38 +51,17 @@ pub fn cast_from_string(
...
-            Ok((builder.build(size), Some(bitmap.into())))
-        }
-        TypeID::DateTime64 => {
+        TypeID::DateTime => {
             let mut builder = ColumnBuilder::<i64>::with_capacity(size);
-            let datetime = data_type.as_any().downcast_ref::<DateTime64Type>().unwrap();
+            let datetime = data_type.as_any().downcast_ref::<DateTimeType>().unwrap();
 
             for (row, v) in str_column.iter().enumerate() {
+                // match string_to_datetime(v) {
```
Comment (b41sh) [inline common/functions/src/scalars/expressions/cast_from_string.rs:59]: unused comment.

**Comment by sundy-li** [comment]
> @dantengsky @sundy-li it seems that we encounter the compatible problem with `meta service` and `fuse engine`.Because meta service and fuse engine have persisted the `type` info.

Will take about this today.

**Comment by dantengsky** [comment]
> Will talk about this today.

roger

File: common/functions/src/scalars/expressions/cast_from_string.rs (hunk 3 above)
Comment (Veeupup) [reply common/functions/src/scalars/expressions/cast_from_string.rs]: resloved

**Comment by Veeupup** [comment]
@sundy-li @b41sh  test has been fixed.

**Review by BohuTANG — APPROVED** [review]
:+1:

File: tests/suites/0_stateless/02_function/02_0012_function_datetimes.sql [hunk 4]
```diff
@@ -187,19 +187,19 @@ select '===EXTRACT===';
...
-insert into ts values(now(), now(), today(), today());
-select toDateTime64(a) = toDateTime64(b),  toDateTime32(a) = toDateTime(b),  toDate32(c) = toDate32(d), toDate(c) = toDate(d) from ts;
+create table ts(a DateTime_3, b DateTime, c Date);
+insert into ts values(now(), now(), today());
+select toDateTime(a) = toDateTime(b) from ts;
 drop table if exists ts;
 
 
-create table t(d32 datetime32, d64 datetime64);
+create table t(d32 datetime, d64 datetime_3);
```
Comment (BohuTANG) [inline tests/suites/0_stateless/02_function/02_0012_function_datetimes.sql:202]: `datetime_3` data type is wired here, do we have plan to change it? @Veeupup @sundy-li

Comment (BohuTANG) [reply tests/suites/0_stateless/02_function/02_0012_function_datetimes.sql:202]: Ok, I will change all the docs about date&time after it ready.

Comment (Veeupup) [reply tests/suites/0_stateless/02_function/02_0012_function_datetimes.sql:202]: @BohuTANG after #4982 merged, `date/datetime` is all ready.

**Pull request merged**